pyquil/api/_quantum_computer.py

```diff
@@ -681,33 +717,60 @@ def local_qvm() -> Iterator[Tuple[subprocess.Popen, subprocess.Popen]]:
 
     >>> from pyquil import get_qc, Program
     >>> from pyquil.gates import CNOT, Z
-    >>> from pyquil.api import local_qvm
+    >>> from pyquil.api import local_forest_runtime
     >>>
     >>> qvm = get_qc('9q-square-qvm')
     >>> prog = Program(Z(0), CNOT(0, 1))
     >>>
-    >>> with local_qvm():
+    >>> with local_forest_runtime():
     >>>     results = qvm.run_and_measure(prog, trials=10)
 
+    :param host: Host on which `qvm` and `quilc` should listen on.
+    :param qvm_port: Port which should be used by `qvm`.
+    :param quilc_port: Port which should be used by `quilc`.
+    :param use_protoquil: Restrict input/output to protoquil.
+
+    .. warning::
+        If ``use_protoquil`` is set to ``True`` language features you need
+        may be disabled. Please use it with caution.
+
     :raises: FileNotFoundError: If either executable is not installed.
     """
-    # Enter. Acquire resource
-    qvm = subprocess.Popen(['qvm', '-S'],
-                           stdout=subprocess.PIPE,
-                           stderr=subprocess.PIPE)
 
-    quilc = subprocess.Popen(['quilc', '-RP'],
-                             stdout=subprocess.PIPE,
-                             stderr=subprocess.PIPE)
+    qvm = None
+    quilc = None
+
+    if _port_used(qvm_port):
+        print(("Unable to start QVM, since the QVM "
+               "port {} is already in use.").format(qvm_port), file=sys.stderr)
+    else:
+        qvm_cmd = ['qvm', '-S', '--host', host, '-p', str(qvm_port)]
+        qvm = subprocess.Popen(qvm_cmd,
```

Comment:
Both qvm and quilc have a non-zero startup time (on the order of 1s). If `subprocess.Popen` does not block until the process completes, then we could have the situation where the user makes a request to the servers before they are truly available.

I suggest we call `wait()` on both `qvm` and `quilc`.

```
In [10]: %time p = subprocess.Popen("sleep 20", shell=True)
CPU times: user 435 µs, sys: 3.11 ms, total: 3.55 ms
Wall time: 3.06 ms

In [11]: %time p.wait()
CPU times: user 1.91 ms, sys: 1.65 ms, total: 3.56 ms
Wall time: 19.3 s
Out[11]: 0
```

Reply:
For what it's worth: unless pyQuil is explicitly planning to get upset about a missing / timing-out quilc (as in `QuantumComputer` initialization), it is OK to send an RPCQ client message before there's a waiting RPCQ server. The ZMQ client will hold on to the message 'til there's a server ready to receive it, and the user won't notice anything but blocking.

Reply:
That's good to know! I'm not sure how qvm will react however. If ZMQ will block anyway, might as well do that upfront here for completeness' sake.

Reply:
I might be misunderstanding, but `Popen.wait` blocks until the process terminates, unless you pass it a `timeout` arg. If we block on a wait call, we can't return to the caller.

You could always pass a timeout as a way to try and give the process some time to startup, but you still don't really know if it's up and ready to receive requests unless you actually send a request to the port, similar to @sauercrowd's previous suggestion.

Reply:
Yeah: my comment doesn't apply to the QVM. Python `requests` will get upset that there's no listening HTTP server.

Reply:
@appleby I was thinking that qvm was going to daemonize itself after it finished setting up. I dunno why I was thinking that. The issue remains.

Reply:
Can we wait for qvm/quilc in subprocess? Won't they keep on blocking?

I implemented sth a couple of days ago which polls until until qvm and quilc are available, happy to implement it that way here if you think that makes sense

https://github.com/sauercrowd/rqubit/blob/0b81231d75c86cb5fb0d936b37e9b6cc21ecce4d/rqubit/quomputer.py#L50

Reply:
Darn. This is messier than I was expecting. 

@sauercrowd looks like we will have to do some polling like in your example code. Unless @appleby @ecpeterson have any other suggestions, then let's add a `poll_timeout=2` and `poll_interval=0.2`, and check whether the servers are available within `poll_timeout`.